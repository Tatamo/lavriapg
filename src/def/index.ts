export * from "./language";
export * from "./parsingtable";
export * from "./token";
