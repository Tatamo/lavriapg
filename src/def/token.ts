/**
 * トークン名
 */
export type Token = string | symbol;

/**
 * トークン化された入力
 * トークン名と、字句規則にマッチした元々の入力
 */
export type TokenizedInput = { token: Token, value: string };

/**
 * 入力の終端を表す終端記号名
 * @type {symbol}
 */
export const SYMBOL_EOF: Token = Symbol("EOF");

/**
 * `S' -> S $` (Sは開始記号)となるような非終端記号S'を表す非終端記号名
 * @type {symbol}
 */
export const SYMBOL_SYNTAX: Token = Symbol("S'");
