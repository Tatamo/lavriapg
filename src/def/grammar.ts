import {Token} from "./token";
import {ILexer} from "../lexer/lexer";

export interface LexDefinitionSection {
	token: Token | null;
	pattern: string | RegExp;
	priority?: number;
	callback?: (lexer: ILexer, token: string | null, value: string, index: number) => any;
}

export type LexDefinitions = Array<LexDefinitionSection>;

export interface SyntaxDefinitionSection {
	ltoken: Token;
	pattern: Array<Token>;
}

export type SyntaxDefinitions = Array<SyntaxDefinitionSection>;

export interface GrammarDefinition {
	lex: LexDefinitions;
	syntax: SyntaxDefinitions;
	start_symbol: Token;
}
