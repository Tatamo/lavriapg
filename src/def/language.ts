import {Token} from "./token";
import {ILexer, LexController} from "../lexer/lexer";

/**
 * 字句解析器の状態を区別するためのラベル型
 */
export type LexStateLabel = string;

/**
 * デフォルトの字句解析器の状態
 */
export const default_lex_state = "default";

/**
 * 字句解析器に与える状態
 */
export interface LexState {
	label: LexStateLabel;
	is_exclusive?: boolean;
}

/**
 * 字句規則マッチ時に呼び出されるコールバック
 */
export type LexCallback = (token: string | null, value: string, lex: LexController) => [string | null, any] | { token: string | null, value: any } | string | null;

/**
 * 単一の字句ルール
 */
// TODO: tokenはlabelに名称変更してもよい？
export interface LexRule {
	token: Token | null;
	pattern: string | RegExp;
	state?: Array<LexStateLabel>;
	is_disabled?: boolean;
	priority?: number;
	callback?: LexCallback;
}

/**
 * 字句規則
 */
export interface LexDefinition {
	rules: Array<LexRule>;
	states?: Array<LexState>;
	default_callback?: LexCallback;
}

/**
 * 構文のreduce時に呼び出されるコールバック
 */
export type GrammarCallback = (children: Array<any>, token: string, lexer: ILexer) => any;

/**
 * 単一の構文ルール
 */
export interface GrammarRule {
	ltoken: Token;
	pattern: Array<Token>;
	callback?: GrammarCallback;
}

/**
 * 構文規則
 */
export interface GrammarDefinition {
	rules: Array<GrammarRule>;
	start_symbol: Token;
	default_callback?: GrammarCallback;
}

/**
 * 言語定義
 */
export interface Language {
	lex: LexDefinition;
	grammar: GrammarDefinition;
}
