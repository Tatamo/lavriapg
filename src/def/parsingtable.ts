import {Token} from "./token";

/**
 * Shiftオペレーション
 */
export type ShiftOperation = { type: "shift", to: number };

/**
 * Reduceオペレーション
 */
export type ReduceOperation = { type: "reduce", grammar_id: number };

/**
 * Shift/Reduceコンフリクト
 */
export type ConflictedOperation = { type: "conflict", shift_to: Array<number>, reduce_grammar: Array<number> };

/**
 * Acceptオペレーション
 */
export type AcceptOperation = { type: "accept" };

/**
 * Gotoオペレーション
 */
export type GotoOperation = { type: "goto", to: number };

/**
 * 構文解析器の実行する命令群
 */
export type ParsingOperation = ShiftOperation | ReduceOperation | ConflictedOperation | AcceptOperation | GotoOperation;

/**
 * 構文解析表
 */
export type ParsingTable = Array<Map<Token, ParsingOperation>>;
