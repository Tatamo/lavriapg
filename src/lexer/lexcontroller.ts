import {DEFAULT_LEX_STATE, Language, LexDefinition, LexRule, LexState, LexStateLabel} from "../def/language";

export type LexRuleLabel = string;

class LexRuleManager {
	// private states: { states: Map<LexStateLabel, LexState>, index: Map<LexStateLabel, Set<number>>, inheritance: Map<LexStateLabel, LexStateLabel> };
	private states: Map<LexStateLabel, { state: LexState, index: Set<number>, inheritance: LexStateLabel | null }>;
	private rules: { rules: Array<LexRule | undefined>, labels: Map<LexRuleLabel, number> };
	// 各ルールに一意なidを割り当てるためのカウンタ
	private id_counter: number;
	// ルールの削除によって割り当てがなくなったid
	private free_ids: Array<number>;
	constructor(language: Language) {
		const lex = language.lex;
		this.id_counter = 0;
		this.free_ids = [];

		// initialize lex states map
		this.states = new Map();
		// もしlexの定義内にデフォルト状態の記述があっても上書きされるだけなので問題ない
		this.addState({label: DEFAULT_LEX_STATE, is_exclusive: false});
		if (lex.states !== undefined) {
			for (const state of lex.states) {
				this.addState(state);
			}
		}

		// initialize lex rules
		this.rules = {rules: [], labels: new Map()};

		for (const rule of lex.rules.map((r) => LexRuleManager.formatLexRule(r))) {
			this.rules.rules[this.id_counter] = rule;
			// 状態ごとにインデックスを張る
			for (const state of rule.state!) {
				// TODO: statesに入れる
				if (!this.states.has(state)) {
					this.addState(state);
				}
				this.states.get(state)!.index.add(this.id_counter);
			}
			this.id_counter += 1;
		}
	}
	addState(label: LexStateLabel): boolean;
	addState(state: LexState): boolean;
	addState(s: LexStateLabel | LexState): boolean {
		let state: LexState;
		if (typeof s === "object") {
			state = s;
		}
		else {
			state = {label: s};
		}
		state = LexRuleManager.formatLexState(state);
		if (this.states.has(state.label)) {
			return false;
		}
		const inheritance: LexStateLabel | null = state.is_exclusive ? null : DEFAULT_LEX_STATE;
		this.states.set(state.label, {state, index: new Set(), inheritance});
		// ループチェック
		if (inheritance !== null) {
			let flg_loop = false;
			let parent = this.states.get(inheritance);
			while (parent !== undefined && parent.inheritance !== null) {
				// 状態を追加するたびにチェックするので、自身にたどりつかないことを調べればよい
				if (parent.inheritance === state.label) {
					flg_loop = true;
					break;
				}
				parent = this.states.get(parent.inheritance);
			}
			if (flg_loop) this.states.get(state.label)!.inheritance = null;
		}
		return true;
	}
	// TODO: パフォーマンス改善
	getRulesItr(state: LexStateLabel): IterableIterator<LexRule> {
		// そんな状態はない
		if (!this.states.has(state)) return [][Symbol.iterator]();

		// 継承を加味
		let result: Array<number> = [];
		let s = this.states.get(state);
		while (s !== undefined) {
			result = result.concat([...s.index]);
			if (s.inheritance === null) break;
			s = this.states.get(s.inheritance);
		}

		return (function* (self, itr) {
			for (const id of itr) {
				if (self.rules.rules[id] !== undefined) yield self.rules.rules[id]!;
			}
		})(this, new Set(result)[Symbol.iterator]());
	}
	addRule(label: LexRuleLabel, rule: LexRule): void {
		// 同名の既存ルールを破棄
		this.removeRule(label);

		const formatted_rule = LexRuleManager.formatLexRule(rule);

		const id = this.free_ids.length > 0 ? this.free_ids.pop()! : this.id_counter++;
		this.rules.rules[id] = formatted_rule;
		this.rules.labels.set(label, id);
		for (const state of formatted_rule.state!) {
			if (!this.states.has(state)) this.addState(state);
			this.states.get(state)!.index.add(id);
		}
	}
	removeRule(label: LexRuleLabel): LexRule | null {
		if (!this.rules.labels.has(label)) {
			return null;
		}
		const id = this.rules.labels.get(label)!;
		const rule = this.rules.rules[id];
		if (rule === undefined) return null;

		for (const state of rule.state!) {
			if (this.states.has(state)) {
				this.states.get(state)!.index.delete(id);
			}
		}
		this.rules.rules[id] = undefined;
		this.free_ids.push(id);
		return rule;
	}
	static formatLexState(state: LexState): LexState {
		return {
			label: state.label,
			is_exclusive: state.is_exclusive !== undefined ? state.is_exclusive : false
		};
	}
	static formatLexRule(rule: LexRule): LexRule {
		// clone rule
		const result = {...rule};
		if (result.is_disabled === undefined) result.is_disabled = false;
		// 状態指定を省略された場合はデフォルト状態のみとする
		if (result.state === undefined) result.state = [DEFAULT_LEX_STATE];
		// 正規表現を字句解析に適した形に整形
		if (result.pattern instanceof RegExp) {
			result.pattern = LexRuleManager.formatRegExp(result.pattern);
		}
		return result;
	}
	private static formatRegExp(pattern: RegExp): RegExp {
		// フラグを整形する
		let flags: string = "";
		// gフラグは邪魔なので取り除く
		// i,m,uフラグがあれば維持する
		if (pattern.ignoreCase) {
			flags += "i";
		}
		if (pattern.multiline) {
			flags += "m";
		}
		if (pattern.unicode) {
			flags += "u";
		}
		// yフラグは必ずつける
		flags += "y";
		return new RegExp(pattern, flags);
	}
}

export class LexController {
	private _lex: LexDefinition;
	private _current_state: LexStateLabel;
	private _state_stack: Array<LexStateLabel>;
	private _rules: LexRuleManager;
	constructor(language: Language) {
		this._lex = language.lex;
		this._current_state = DEFAULT_LEX_STATE;
		this._state_stack = [];
		this._rules = new LexRuleManager(language);
	}
	onBegin(): void {
		if (this._lex.begin_callback !== undefined) this._lex.begin_callback(this);
	}
	onEnd(): void {
		if (this._lex.end_callback !== undefined) this._lex.end_callback(this);
	}
	getRulesItr(): IterableIterator<LexRule> {
		return this._rules.getRulesItr(this._current_state);
	}
	addRule(label: string, rule: LexRule): void {
		this._rules.addRule(label, rule);
	}
	removeRule(label: string): LexRule | null {
		return this._rules.removeRule(label);
	}
	getCurrentState(): LexStateLabel {
		return this._current_state;
	}
	jumpState(state: LexStateLabel): void {
		this._current_state = state;
	}
	callState(state: LexStateLabel): void {
		this._state_stack.push(this._current_state);
		this._current_state = state;
	}
	returnState(): LexStateLabel | undefined {
		const pop = this._state_stack.pop();
		if (pop === undefined) this._current_state = DEFAULT_LEX_STATE;
		else this._current_state = pop;
		return pop;
	}
}
