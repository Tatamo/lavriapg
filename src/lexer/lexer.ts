import {Language, LexRule} from "../def/language";
import {SYMBOL_EOF, TokenizedInput} from "../def/token";
import {LexController} from "./lexcontroller";

/**
 * 字句解析器用のinterface
 *
 * TODO: 要改善
 */
export interface ILexer {
	exec(input: string): Array<TokenizedInput>;
}

/**
 * 字句解析器
 * 入力を受け取ってトークン化する
 */
export class Lexer implements ILexer {
	constructor(private language: Language) {
		// do nothing
	}
	exec(input: string): Array<TokenizedInput> {
		const result: Array<TokenizedInput> = [];
		let next_index = 0;
		const controller = new LexController(this.language);
		controller.onBegin();
		while (next_index < input.length) {
			// 念の為undefined対策
			// const current_rules = this.rules.has(controller.getCurrentState()) ? this.rules.get(controller.getCurrentState())! : [];
			const current_rules = controller.getRulesItr();
			const {rule, matched} = Lexer.match(current_rules, input, next_index);
			if (rule === null) {
				// マッチする規則がなかった
				throw new Error("no pattern matched");
			}
			else {
				let token = rule.token;
				let value = matched;
				// コールバック呼び出し
				if (typeof rule.token !== "symbol" && rule.callback !== undefined) {
					const callback_result = rule.callback(matched, rule.token, controller);
					if (callback_result === null) {
						token = null;
					}
					else if (typeof callback_result === "string") {
						token = callback_result;
					}
					else if (Array.isArray(callback_result)) {
						token = callback_result[0];
						value = callback_result[1];
					}
					else if (callback_result !== undefined) {
						token = callback_result.token;
						value = callback_result.value;
					}
					// callback_result === undefinedなら何もしない
				}
				// tokenがnullなら処理を飛ばす
				if (token !== null) {
					result.push({token: token, value: value});
				}
				// 読む位置を進める
				next_index += matched.length;
			}
		}
		result.push({token: SYMBOL_EOF, value: ""});
		controller.onEnd();
		return result;
	}
	private static match(rules: Iterable<LexRule>, input: string, next_index: number): { rule: LexRule | null, matched: string } {
		let result_matched: string = "";
		let result_rule: LexRule | null = null;
		let result_priority: number | null = null;
		for (const rule of rules) {
			let match = "";
			if (typeof rule.pattern === "string") {
				const tmp_next_index = next_index + rule.pattern.length;
				if (input.substring(next_index, tmp_next_index) != rule.pattern) continue; // マッチしない
				// マッチした文字列の末尾が\wで、その直後の文字が\wの場合はスキップ
				if (tmp_next_index < input.length && /\w/.test(rule.pattern.substring(0, 1)) && /\w/.test(input[tmp_next_index])) continue;
				match = rule.pattern;
			}
			else {
				// pattern: RegExp
				rule.pattern.lastIndex = next_index;
				const m = rule.pattern.exec(input);
				if (m === null) continue; // マッチ失敗
				match = m[0];
			}
			// 同じ優先度の場合、最長マッチまたは出現順(match_priorityで設定)
			const priority = rule.priority !== undefined ? rule.priority : 0;
			if (result_priority === null ||
				priority > result_priority ||
				priority === result_priority && match.length > result_matched.length) {
				result_matched = match;
				result_rule = rule;
				result_priority = priority;
			}
		}
		return {rule: result_rule, matched: result_matched};
	}
}
