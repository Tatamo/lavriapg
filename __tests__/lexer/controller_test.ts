import {test_dynamic_lexrules_language, test_lexstate_language} from "../data/sample_language";
import {SYMBOL_EOF} from "../../src/def/token";
import {Lexer} from "../../src/lexer/lexer";

describe("lex state test", () => {
	test("nested states", () => {
		const lexer = new Lexer(test_lexstate_language);
		expect(lexer.exec("${$($123)$}")).toEqual([
			{token: "INVALID", value: "$"},
			{token: "LBRACE", value: "{"},
			{token: "DOLLAR", value: "$"},
			{token: "LPAREN", value: "("},
			{token: "INVALID", value: "$"},
			{token: "NUMBER", value: "123"},
			{token: "RPAREN", value: ")"},
			{token: "DOLLAR", value: "$"},
			{token: "RBRACE", value: "}"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
	test("rule of non-default state", () => {
		const lexer = new Lexer(test_lexstate_language);
		expect(lexer.exec("123")).not.toEqual([
			{token: "NUMBER", value: "123"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
	test("reset state after process", () => {
		const lexer = new Lexer(test_lexstate_language);
		expect(lexer.exec("{(")).toEqual([
			{token: "LBRACE", value: "{"},
			{token: "LPAREN", value: "("},
			{token: SYMBOL_EOF, value: ""}
		]);
		expect(lexer.exec(")}")).toEqual([
			{token: "INVALID", value: ")"},
			{token: "INVALID", value: "}"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
	test("exclusive state", () => {
		const lexer = new Lexer(test_lexstate_language);
		expect(lexer.exec("(+a{*)")).toEqual([
			{token: "LPAREN", value: "("},
			{token: "PLUS", value: "+"},
			{token: "INVALID", value: "a"},
			{token: "INVALID", value: "{"},
			{token: "ASTERISK", value: "*"},
			{token: "RPAREN", value: ")"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
	test("non-exclusive state", () => {
		const lexer = new Lexer(test_lexstate_language);
		expect(lexer.exec("${$+a*}")).toEqual([
			{token: "INVALID", value: "$"},
			{token: "LBRACE", value: "{"},
			{token: "DOLLAR", value: "$"},
			{token: "PLUS", value: "+"},
			{token: "ID", value: "a"},
			{token: "INVALID", value: "*"},
			{token: "RBRACE", value: "}"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
});

describe("dynamic lex rules test", () => {
	test("adding and removing rules", () => {
		const lexer = new Lexer(test_dynamic_lexrules_language);
		expect(lexer.exec("%%{}%}%%}%%")).toEqual([
			{token: "LNEST", value: "%%{"},
			{token: "INVALID", value: "}"},
			{token: "INVALID", value: "%"},
			{token: "RNEST", value: "}%%"},
			{token: "INVALID", value: "}"},
			{token: "INVALID", value: "%"},
			{token: "INVALID", value: "%"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
});
