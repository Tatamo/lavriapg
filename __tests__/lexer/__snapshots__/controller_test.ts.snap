// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`begin/end callbacks test using variables 1`] = `
Array [
  Object {
    "token": "A",
    "value": "0",
  },
  Object {
    "token": "A",
    "value": "1",
  },
  Object {
    "token": "A",
    "value": "2",
  },
  Object {
    "token": "A",
    "value": "3",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`begin/end callbacks test using variables 2`] = `
Array [
  Object {
    "token": "A",
    "value": "0",
  },
  Object {
    "token": "A",
    "value": "1",
  },
  Object {
    "token": "A",
    "value": "2",
  },
  Object {
    "token": "A",
    "value": "3",
  },
  Object {
    "token": "A",
    "value": "4",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`dynamic lex rules test adding and removing rules 1`] = `
Array [
  Object {
    "token": "LNEST",
    "value": "%%{",
  },
  Object {
    "token": "INVALID",
    "value": "}",
  },
  Object {
    "token": "INVALID",
    "value": "%",
  },
  Object {
    "token": "RNEST",
    "value": "}%%",
  },
  Object {
    "token": "INVALID",
    "value": "}",
  },
  Object {
    "token": "INVALID",
    "value": "%",
  },
  Object {
    "token": "INVALID",
    "value": "%",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`lex state test exclusive state 1`] = `
Array [
  Object {
    "token": "LPAREN",
    "value": "(",
  },
  Object {
    "token": "PLUS",
    "value": "+",
  },
  Object {
    "token": "INVALID",
    "value": "a",
  },
  Object {
    "token": "INVALID",
    "value": "{",
  },
  Object {
    "token": "ASTERISK",
    "value": "*",
  },
  Object {
    "token": "RPAREN",
    "value": ")",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`lex state test nested states 1`] = `
Array [
  Object {
    "token": "INVALID",
    "value": "$",
  },
  Object {
    "token": "LBRACE",
    "value": "{",
  },
  Object {
    "token": "DOLLAR",
    "value": "$",
  },
  Object {
    "token": "LPAREN",
    "value": "(",
  },
  Object {
    "token": "INVALID",
    "value": "$",
  },
  Object {
    "token": "NUMBER",
    "value": "123",
  },
  Object {
    "token": "RPAREN",
    "value": ")",
  },
  Object {
    "token": "DOLLAR",
    "value": "$",
  },
  Object {
    "token": "RBRACE",
    "value": "}",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`lex state test non-exclusive state 1`] = `
Array [
  Object {
    "token": "INVALID",
    "value": "$",
  },
  Object {
    "token": "LBRACE",
    "value": "{",
  },
  Object {
    "token": "DOLLAR",
    "value": "$",
  },
  Object {
    "token": "PLUS",
    "value": "+",
  },
  Object {
    "token": "ID",
    "value": "a",
  },
  Object {
    "token": "INVALID",
    "value": "*",
  },
  Object {
    "token": "RBRACE",
    "value": "}",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`lex state test reset state after process 1`] = `
Array [
  Object {
    "token": "LBRACE",
    "value": "{",
  },
  Object {
    "token": "LPAREN",
    "value": "(",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`lex state test reset state after process 2`] = `
Array [
  Object {
    "token": "INVALID",
    "value": ")",
  },
  Object {
    "token": "INVALID",
    "value": "}",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;
